### PR TITLE
Save revert should disable dirty fields

### PR DESCRIPTION
**Pull request by oehm-smith**
GEOD-417

Also stopped the form from being saved if it is invalid.

DateTime not modified like others due to bug - I've GEOD-426 open for this.

### CODE REVIEW

File: src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts [hunk 1]
```diff
@@ -27,4 +27,38 @@ export abstract class AbstractGnssControls {
         }
         return errString;
     }
+
+    protected isDirty(controlId: string): boolean {
+        let control: AbstractControl;
+        let dirty: boolean = true;
+        // Force an error as this field should have been set
```
Comment (lbodor) [inline src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts]: Is this comment relevant. Where is the error being forced?

File: src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts [hunk 2]
```diff
@@ -27,4 +27,38 @@ export abstract class AbstractGnssControls {
...
+        if (!control) {
+            valid = true;
+        } else {
+            valid = control.valid;
+        }
+        console.warn('  isValid Control "' + controlId + '": '+ valid + ' - ', control);
+        return valid;
+    }
+
+    protected isInValid(controlId: string): boolean {
```
Comment (lbodor) [inline src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts]: Spelling: invalid is one word.

Comment (lbodor) [inline src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts]: `isValid` would be even better.

File: src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts [hunk 3]
```diff
@@ -27,4 +27,38 @@ export abstract class AbstractGnssControls {
...
+        let valid: boolean = this.isValid(controlId);
+        // console.debug(`isInValid for ${controlId} - ${! valid}`);
```
Comment (lbodor) [inline src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts]: Remove commented out code.

File: src/client/app/shared/dynamic-form-fields/datetime-input.component.html [hunk 4]
```diff
@@ -1,17 +1,17 @@
-<div class="form-group">
-    <label class="col-md-4 col-sm-4 col-xs-5 col-xxs-12 control-label" [ngClass]="{required: isRequired()}">
-        <ng-content></ng-content>
-    </label>
+<div class="form-group"> <!-- [formGroup]="form"-->
```
Comment (lbodor) [inline src/client/app/shared/dynamic-form-fields/datetime-input.component.html]: Remove commented out code.

File: src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts (hunk 1 above)
Comment (oehm-smith) [inline src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts]: Removed them.

File: src/client/app/shared/dynamic-form-fields/datetime-input.component.html [hunk 5]
```diff
@@ -1,17 +1,17 @@
...
-            <div class="input-group">
-                <input type="text" maxlength="24"
-                       class="form-control z-lowest"
-                       [(ngModel)]="datetimeDisplay"
-                       [ngClass]="{'ng-dirty': hasChanges, 'ng-invalid': invalidDatetime }"
-                       (change)="updateCalendar()"
-                       (focus)="miscUtils.scrollIntoView($event, showDatetimePicker); showDatetimePicker = true;"/>
+    <div>
+      <div class="input-group"> <!-- [ngClass]="{'gnss-dirty': hasChanges, 'gnss-invalid':isInValid(controlName)}">   -- Or invalidDatetime -->
+               <!--[formControlName]="controlName"-->
```
Comment (lbodor) [inline src/client/app/shared/dynamic-form-fields/datetime-input.component.html]: Remove commented out code.

File: src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts (hunk 2 above)
Comment (oehm-smith) [inline src/client/app/shared/dynamic-form-fields/abstract-gnss-controls.ts]: Historical and meant to fix.  WIll do.

File: src/client/app/site-info/site-info.component.ts [hunk 6]
```diff
@@ -182,7 +187,13 @@ export class SiteInfoComponent implements OnInit, OnDestroy {
 
         this.removeDeletedItems();
 
+<<<<<<< HEAD
         this.dialogService.confirmSaveDialog(
+=======
+        let that = this;
+
+        this.dialogService.confirmSaveDialog(diffMsg,
+>>>>>>> GEOD-417 - Fix so a save or revert makes everything pristine
```
Comment (lbodor) [inline src/client/app/site-info/site-info.component.ts]: Merge markers.

File: src/client/app/site-info/site-info.component.ts [hunk 7]
```diff
@@ -192,6 +203,8 @@ export class SiteInfoComponent implements OnInit, OnDestroy {
                     (responseJson: any) => {
                         //if (form)form.pristine = true;  // Note: pristine has no setter method in ng2-form!
```
Comment (lbodor) [inline src/client/app/site-info/site-info.component.ts]: Remove commented out code.

File: src/client/app/site-info/site-info.component.ts [hunk 8]
```diff
@@ -288,6 +301,21 @@ export class SiteInfoComponent implements OnInit, OnDestroy {
             if (this.siteInfoForm.dirty) {
                 this.siteLogService.sendFormModifiedStateMessage(true);
                 console.log('form dirty - yes: ', value);
+                // console.log('  and siteLogModel: ', this.siteLogModel);
+                // console.log('  and siteLogOrigin: ', this.siteLogOrigin);
```
Comment (lbodor) [inline src/client/app/site-info/site-info.component.ts]: Remove commented out code.

File: src/client/app/site-info/site-info.component.ts [hunk 9]
```diff
@@ -288,6 +301,21 @@ export class SiteInfoComponent implements OnInit, OnDestroy {
...
+            }
+        });
+
+        this.siteInfoForm.statusChanges.debounceTime(500).subscribe((value: any) => {
+            console.debug('form status change: ', this.siteInfoForm);
+            if (this.siteInfoForm.dirty) {
+                this.siteLogService.sendFormModifiedStateMessage(true);
+                console.log('form dirty - yes: ', value);
+                // console.log('  and siteLogModel: ', this.siteLogModel);
+                // console.log('  and siteLogOrigin: ', this.siteLogOrigin);
```
Comment (lbodor) [inline src/client/app/site-info/site-info.component.ts]: Remove commented out code.

File: src/client/app/shared/dynamic-form-fields/datetime-input.component.html (hunk 4 above)
Comment (oehm-smith) [inline src/client/app/shared/dynamic-form-fields/datetime-input.component.html]: No that is what I need to insert.  I have GEOD-426 to fix this.

**Pull request merged**